text/0864-deprecate-ember-a.md

```diff
@@ -0,0 +1,94 @@
+
+Along with it's usefulness `Ember.A()` has a significant API gotcha in that it doesn't return a new instance of the array-like object it is passed, it modifies that object and returns is such that `A(array) === array` this can create the dreaded spooky-action-at-a-distance effect where suddenly an object which was `Array` can become an `Ember.NativeArray`. An example of this is using `{{includes "Zoey" this.mascots}}` from [ember-composable-helpers](https://github.com/DockYard/ember-composable-helpers#includes) will modify `this.mascots` and add `lastObject` to the API. Re-ordering the template code at a later date will result in a mysterious failure that can be extremely frustrating to understand.
+
+```js
+export default class WeirdArray extends Component {
+  mascots = ['Tomster', 'Zoey'];
+
+  get lastMascot() {
+    return this.mascots.lastObject;
+  }
+
+  <template>
+    {{#if (includes "Zoey" this.mascots)}}
+      Zoey Rules!
+    {{/if}}
+    Last Mascot: {{this.lastMascot}}
+  </template>
+}
+```
+
+Deprecating `Ember.A()` will provide a signal to the addon community to move away from both array prototype extensions and array reactivity which requires using the Emberism `pushObject`/`removeObject` to track updates. 
+
+Now that array prototype extensions have been deprecated it is quite possible that usage of `Ember.A()` will increase significantly as a defensive coding strategy to maintain access to the family of methods Ember adds to all arrays. This should be immediately discouraged as `Ember.A()` is not a long term viable solution and other better alternatives exist.
+
+## Transition Path
+
+### Replacement APIs
+
+As discussed in #848 better alternatives exist for the APIs provided by `Ember.A()`:
+
+> For convenient methods like `filterBy`, `compact`, `sortBy` etc., the replacement functionalities already exist either through native array methods or utility libraries like [lodash](https://lodash.com), [Ramda](https://ramdajs.com), etc.
+
+>For mutation methods (like `pushObject`, `removeObject`) or observable properties (like `firstObject`, `lastObject`) participating in the Ember classic reactivity system, the replacement functionalities also already exist in the form of immutable update style with tracked properties like `@tracked someArray = []`, or through utilizing `TrackedArray` from `tracked-built-ins`.
+
+### Clearly Define Idiomatic Path
+
+Deprecating `Ember.A()` will provide a clear signal to move away from this construct and rely on these more robust modern alternatives.
+
+### Transitional Feature Flags
+
+Transition away from `Ember.A()` should be aided by a pair of feature flags:
+
+#### Wrap Passed Value
+
+Instead of returning the passed value `Ember.A()` would return a native proxy to wrap the original object without modifying it. This would remove the most confusing part of the API and allow apps to opt in early and clear any unexpected issues before `Ember.A()` is fully removed.
```

Comment:
I've started doing some of this work in https://github.com/emberjs/ember.js/pull/20245. This change shouldn't actually require a flag since it's how Ember.A claims to work. However, we could add a short-lived flag to make it easier to validate the behavior in some real apps. Once we successfully validated we'd just remove the flag and have it always on. I don't think we'd have to publicize it in this case.

Once we successfully validate this, I'd like to add individual deprecations to the custom methods on the wrapper. That would help us to quickly identify cases where we're actually relying on the Ember.A behavior.

Reply:
I'm not sure if I need to express this in the RFC or if this is just for information / reference. I agree with most of it, though I don't share your faith that no one is relying on this behavior. I think it is probably fair to say that the number of apps *intentionally* relying in this is low, however I think the side effect may have turned into a feature for many.